### PR TITLE
[BEAM-7746] Resolve typing issues in filesystem

### PR DESCRIPTION
**Pull request by chadrik**
In order to properly type some of the complexities around the `filesystem` API I had to address 2 main problems:

- properly type `CompressedFile`
  - add `Compressor` and `Decompressor` protocols
  - make it a `BinaryIO` subclass.  this is necessary to guarantee that all the file-like things that are floating around these modules actually present the same interface.  unfortunately, `BinaryIO` is not a protocol, so it must be subclassed.  this required a few breaking changes:
    - `writeable` renamed to `writable`
    - `closed` method changed to a property
    - `seekable` property changed a method
    - `read` arg changed name
    - `write` arg changed name
-  add  a new `create_buffered()` classmethod to `DownloaderStream` and `UploaderStream`
   - simplifies the creation of buffered wrappers to the classes
   - this was motivated primarily by the desire to consolidate this pattern into one place so that I could handle the unfortunate need to `cast` these types to `BinaryIO`.   see notes on `DownloaderStream.create_buffered()` for why this was necessary.

--- 

This PR makes one significant runtime change that must be vetted.

`s3io.S3IO.open()` originally created a buffered `DownloaderStream` like this:

```python
      return io.BufferedReader(
          DownloaderStream(downloader, mode=mode), 
          buffer_size=read_buffer_size)
```

All other IO clients pass `read_buffer_size` to `DownloaderStream`, like this example from `gcsio`:

```python
      return io.BufferedReader(
          DownloaderStream(downloader, read_buffer_size=read_buffer_size, mode=mode),
          buffer_size=read_buffer_size)
```

In this PR I standardized on the latter behavior in order to simplify `create_buffered`, but I don't know whether this difference in `s3io` was intentional or not, or what the impact of changing it would be. 

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)
XLang | --- | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/)

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

File: sdks/python/apache_beam/io/localfilesystem.py [hunk 1]
```diff
@@ -136,10 +137,17 @@ def _path_open(
...
+    # type: (...) -> BinaryIO
+
     """Helper functions to open a file in the provided mode.
     """
     compression_type = FileSystem._get_compression_type(path, compression_type)
-    raw_file = open(path, mode)
+    if mode == 'r':
+      mode = 'rb'
+    elif mode == 'w':
+      mode = 'wb'
```
Comment (chadrik) [inline sdks/python/apache_beam/io/localfilesystem.py]: I saw this pattern another filesystem so I thought I'd add it here to make it completely obvious that the subsequent cast to `BinaryIO` (from I`O[Any]`) is safe and warranted.

**Comment by chadrik** [comment]
R: @robertwb 
R: @udim

**Comment by chadrik** [comment]
Run Python PreCommit

**Comment by chadrik** [comment]
Run Python PreCommit

**Comment by chadrik** [comment]
The lint job is failing on a file that doesn't exit...

```
************* Module apache_beam.runners.interactive.utils
apache_beam/runners/interactive/utils.py:25:0: W0611: Unused WindowedValue imported from apache_beam.utils.windowed_value (unused-import) 
```

File: sdks/python/apache_beam/io/filesystemio.py [hunk 2]
```diff
@@ -175,10 +190,31 @@ def readall(self):
...
+    # must inherit from it.  We could roll our own BinaryIO Protocol, but the
+    # stubs for io.Buffered* do not have the required 'mode' or 'name' attrs to
+    # meet the protocol (the classes seem to expose 'mode' and 'name'
+    # conditionally, if their underlying io.RawIOBase possess the attributes).
+    # Thus, it is necessary to cast these types to BinaryIO either way.
+    return cast(
+        BinaryIO,
+        io.BufferedReader(
+            cls(downloader, read_buffer_size=read_buffer_size, mode=mode),
+            buffer_size=read_buffer_size))
```
Comment (chadrik) [inline sdks/python/apache_beam/io/filesystemio.py]: should I create separate args for the buffer size of the `DownloaderStream` and the `io.BufferedReader`?   The treatment of these varies between filesystems.

**Comment by robertwb** [comment]
Is this subsumed into the other PR?

**Comment by chadrik** [comment]
No, this is independent.

**Pull request closed without merging**